### ISSUE TITLE
F5a - The connection controller: the state machine

### ISSUE THREAD

**Issue by lalitkarthik** [lalitkarthik/delta-exchange-payoff#38]
## Parent

#33. The connection controller, first half: the state machine.

## What to build

Every adapter's connection is in exactly one of five states — `connecting`, `connected`, `degraded`, `reconnecting`, `stopped` — and every change of state is a `feed.connection` event on the bus carrying from-state, to-state, reason and adapter. A socket that goes quiet for longer than a bounded interval becomes `degraded` without anyone asking; the next message makes it `connected` again; a longer silence or a closed socket makes it `reconnecting`. Driven by the fake adapter's script, all of this is visible from the outside as a sequence of events.

## Concept

A **state machine** is a small set of named states and the only allowed moves between them. Its value is what it forbids: a connection cannot be both connected and reconnecting, and cannot become connected without passing through connecting. **Staleness** is the observation that a socket which is open but silent is not the same as one that is receiving — the operating system will happily keep a dead TCP connection open for minutes. The controller's job is to turn "nothing has arrived" into a state a person can read, rather than leaving it as an absence.

## Why this way

Today the feed reconnects and resubscribes inside its own loop, and nothing outside knows. Health says ok whether or not the venue is speaking. A browser looking at a ladder that stopped cannot tell a quiet market from a dead feed — the plausible-and-wrong failure this project refuses. Making state explicit, and making every transition an event, is what lets #40 put a badge on the ladder and #42 log every transition, without either of them reaching inside the controller. The controller is split from the budget-and-replay half (#39) because the machine is testable on its own with the fake, and the move of reconnect logic out of the feed is a second, riskier change.

## Learn first

- Michael Nygard, *Release It!*, the chapters on stability patterns — Circuit Breaker and Heartbeat are the two this controller is built from.
- The feed module's docstring on why it reconnects the way it does, and what "resubscribe everything" prevents.
- The events LLD from #35: the `feed.connection` and `heartbeat` events this controller emits.
- The fake adapter's script verbs from #36 — the controller's tests are written in them.

## Task

- A controller class wrapping one adapter: the five states, the transition table, and a method per cause (message arrived, socket closed, staleness timer fired, stopped by request). Every transition publishes `feed.connection` and, for now, logs one line through the existing logger.
- Staleness: no message for the degraded interval → `degraded`; no message for the longer reconnect interval → `reconnecting`. Both intervals are configuration. The degraded default is `assumed` 15 s, chosen as three ticker refreshes at `measured` 5001 ms; the ticket re-measures the live feed's longest quiet gap over an hour and records it in the LLD before trusting the default.
- A `heartbeat` event per adapter on a fixed cadence, carrying the age of the last message.
- The controller drives the adapter's stream; reconnect itself (backoff, budget, replay) stays in the adapter for this ticket and moves in #39.
- The controller LLD, including the transition table.

## How you will know

- [x] A test parametrised over the transition table: every allowed move produces one `feed.connection` event with the right from, to and reason; every forbidden move raises.
- [x] Fake script "frames, silence 20 s": events show `connected → degraded` at the degraded interval and nothing else; "frames, silence 20 s, frames": `degraded → connected` on the first frame back.
- [x] Fake script "frames, close": `connected → reconnecting` with reason `closed`.
- [x] Heartbeat events arrive on cadence with a last-message age that grows during silence and resets on a frame.
- [x] Engine started live for one hour: the longest gap between messages is recorded (`measured`, run named) in the LLD beside the chosen default.
- [x] The controller LLD exists with the transition table.

## What to notice

Which transitions never fire in an hour of live running, and which fire more than expected. If `degraded` fires on a quiet Sunday minute, the default is too tight; if it never fires during a real disconnect, too loose. The number is the finding.

## Acceptance criteria

See *How you will know*.

## Blocked by

- #36 — the adapter protocol and the fake that drives these tests.

**Comment by lalitkarthik** [comment]
**Read this before starting.** #36 has landed and the fake adapter you depend on exists.

- **The fake is `engine/tests/fakes/scripted_adapter.py`**, imported as `from fakes.scripted_adapter import ...`. Four verbs — `Frames`, `Close`, `Silence(seconds)`, `Resume` — documented in that module and in `docs/design/lld/adapter.md` §7. Its clock is injected, so `Silence(20.0)` costs no wall-clock time in a test.
- **It deliberately emits no `feed.connection` events.** Those are yours to emit.
- **`Close` is observable** through `closes`, `connections` and `replays` — an empty `replays` snapshot means a reconnect that replayed nothing, which is the silent failure the resubscribe-everything rule exists to prevent.
- **The adapter protocol carries no connection signal today.** You will need to add one; say what you added.
- **Reconnect logic is still inside `feed.DeltaFeed`, untouched and fully tested.** Leave it there — #39 lifts it. This ticket is the state machine only.
- `main.DeltaFeed` reaches the adapter as `feed_factory`, which is the seam the lifespan tests drive.

**Comment by lalitkarthik** [comment]
**Delivered.** Merged to `main` — commits `9368714` and `13d3162`, merge `4624940`.

Five-state machine with all 13 transitions from `hld.md` §3, staleness, heartbeats, `feed.connection`/`heartbeat`/`alert` events, a new `on_connection`/`off_connection` pair on the adapter protocol, `docs/design/lld/controller.md` and a new `docs/design/lld/connection-signal.md`.

Verified independently: **889 passed / 0 failed** on `main` after merge, ruff clean, `tsc` clean, web guard passing, build clean. No assistant attribution.

The transition table is **hand-transcribed from `hld.md` into the test**, not imported from the module's own constant, so it can genuinely disagree with the code. Red-green verified: dropping `degraded→reconnecting` and adding `reconnecting→connected` produced exactly two failures.

## The hour, and what it nearly caught

`tools/measure_quiet_gap.py`, **3,610 s from 2026-09-07T13:59:51Z**, 492 BTC symbols, both channels, **3,808,870 messages**, 3 connections, 0 malformed:

| Run | Window | Messages | **Max gap** | p99 | p95 |
|---|---|---|---|---|---|
| `20260907T135906Z` | 35 s | 25,033 | 0.321 s | 0.011 | 0.002 |
| `20260907T144659Z` | 550 s | 594,024 | 3.355 s | 0.011 | 0.002 |
| `20260907T135951Z` | **3,610 s** | 3,808,870 | **44.785 s** | 0.011 | 0.002 |

**Only the tail moves.** p99, p95 and median are identical across all three; the maximum moves two orders of magnitude. The 35 s run alone would have said 15 s is 47x the worst gap and been confidently wrong — a good argument against short runs generally.

- `degraded_after` = **15 s** is *not* too tight: crossed twice in the hour, both by real interruptions, never by a quiet market. The ticket's "quiet Sunday minute" worry did not materialise.
- `reconnect_after` = **45 s** was **not** crossed — **by 0.215 s**. No margin worth the name. A second hour is worth taking, and it is #39's default to own.

Recorded in `docs/design/quiet-gap.md` (new, since `tools/out/` is gitignored — the tool also used to overwrite its own output every run, so the 35 s result was one re-run from gone; output is now named by start time).

**Honest limits stated by the agent:** the hour was not clean — three connections means two reconnects, and `last_error` records a local `WinError 1236`, so at least one interruption was this machine's rather than Delta's. The recorder does not tag a gap with the connection it spanned, so attributing gaps to reconnects is inference, not measurement. A quiet market, a weekend and ETH all remain unmeasured.

## Review findings, all fixed

A review of `9368714` found two HIGH bugs, both reproduced, both fixed red-green in `13d3162`:

1. **Every real reconnect would have flapped forever.** `_last_message_at` was never rebased on reopen, so after any outage longer than `reconnect_after` the machine reached `connected` and demoted one second later, emitting three spurious `feed.connection` events per cycle — and #40 badges on those, so the badge would have flickered precisely during the incident an operator is watching. **The hour shows this came 0.215 s from firing in production.** Fixed by measuring from the later of the reopen and the last message; the heartbeat still reports the venue's real silence, because a badge claiming "0 s since last message" one second after a reopen is the same failure one layer up.
2. **The staleness watchdog died silently if any consumer raised.** `gather(..., return_exceptions=True)` retrieved the exception, suppressing even asyncio's warning. Reproduced: with a raising consumer the controller sat in `connected` through a silence five times past the reconnect bound and `run()` returned cleanly. Now caught, logged with traceback, counted, and an `alert` after three consecutive failures.

Also fixed: `OPENED` firing on a socket with zero subscriptions; unremovable connection listeners (`off_connection` is the eighth protocol member); and **a test that could not fail** — `test_a_stop_is_not_reported_as_a_drop` asserted something guaranteed by its own control flow, proved vacuous by running the old body against a deliberately broken guard and watching it pass.

## For #39, #40, #41

```
ConnectionSignal.OPENED / .CLOSED          # str Enum
ConnectionListener = Callable[[ConnectionSignal, str], None]
adapter.on_connection(listener)  -> None
adapter.off_connection(listener) -> None   # 8th member, by equality, silent if absent
```

`OPENED` means *up **and** resubscribed **and** non-empty* — an empty registry fires nothing. `transition(to, reason)` is public: #39 drives `reconnecting → stopped` on a spent budget, #41 the commands. `controller.detach()` is idempotent, `run()` detaches on exit, `start()` re-attaches. `DeltaFeed.empty_opens` is new for `/health`. Two new alert codes: `connection_silent`, `poll_failing`.

**One exposure deliberately left:** `transition()` publishes unguarded, so a consumer raising there propagates out of `run()`. Arguably correct signalling to a supervisor, and guarding it would change a contract #39/#41 call — flagged rather than decided.

**Merge note:** #37 renamed `feed.py` → `adapters/delta_socket.py` while this branch was in flight. Seven conflict hunks were resolved by hand — the duplicated decoder helper was dropped in favour of #37's shared `fakes/decoder.py`, and a `_last_spot` field was removed as dead, since #37 made `md.index_quote` once-per-frame rather than once-per-underlying.

**Orchestration error, mine:** two agents ran in this worktree at once — the implementer and the review-fixer — because I spawned the second believing the first had finished. They collided over files. No work was lost, but it caused one agent to misread a half-written file as fabricated and say so before retracting it.
